core/scheduler.js

```diff
@@ -0,0 +1,108 @@
+/*
+  Copyright 2015 Google Inc. All Rights Reserved.
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+var Promise = require('bluebird');
+var assert = require('chai').assert;
+
+var taskQueue = [];
+var maxWaiting = 32;
+var waitCount = 0;
+
+function startPhaseList(phases) {
+  var initPhases = phases
+      .map(function(phase, idx) { return {phase: phase, idx: idx} })
+      .filter(function(phase) { return phase.phase.init !== undefined; });
+
+  var finishedPromiseList = [];
+
+  return Promise.all(initPhases.map(function(phase) {
+    return phase.phase.init(function(stream) {
+      var promise = new Promise(function(resolve, reject) {
+        schedule(phases, phase.idx + 1, stream, resolve);
```

Comment:
Future patch, I think. We probably just need to plumb failure through the returned promises?